File: neo/backend/src/app/api/v1/skills.py
"""Skills API routes."""

from fastapi import APIRouter, Depends, Path, Query
from sqlalchemy.orm import Session

from app.core.error_codes import (
    ERR_BAD_REQUEST,
    ERR_CONFLICT,
    ERR_NOT_FOUND,
    ERR_OK,
)
from app.database import get_db
from app.models.skill import SkillLevel, SkillStatus
from app.schemas.file import (
    FileMetadataCreate,
    FileMetadataUpdate,
)
from app.schemas.skill import (
    SkillCreate,
    SkillUpdate,
)
from app.schemas.skill_version import (
    SkillRollbackRequest,
    SkillVersionCreate,
)
from app.services.skill_service import SkillService

router = APIRouter(prefix="/api/v1/skills", tags=["skills"])


def _make_response(data=None, code: int = ERR_OK, message: str = "ok") -> dict:
    """Helper to create standard response."""
    return {
        "code": code,
        "message": message,
        "data": data,
        "traceId": "",
        "timestamp": 0,
    }


def _make_error_response(code: int, message: str) -> dict:
    """Helper to create error response."""
    return _make_response(None, code, message)


def _skill_to_detail(skill, versions=None) -> dict:
    """Convert Skill model to detail response."""
    current_version = None
    if versions:
        latest = versions[0] if versions else None
        if latest:
            current_version = {
                "id": latest.id,
                "version": latest.version,
                "comment": latest.comment,
                "created_at": latest.created_at,
            }

    draft_snapshot = None
    if skill.draft_snapshot:
        draft_snapshot = [
            {"file_metadata_id": item["file_metadata_id"], "file_id": item["file_id"]} for item in skill.draft_snapshot
        ]

    return {
        "id": skill.id,
        "code": skill.code,
        "name": skill.name,
        "level": skill.level.value,
        "tags": skill.tags or [],
        "status": skill.status.value,
        "draft_snapshot": draft_snapshot,
        "current_version": current_version,
        "create_user_id": skill.create_user_id,
        "created_at": skill.created_at.isoformat(),
        "updated_at": skill.updated_at.isoformat(),
    }


@router.post("", response_model=dict)
async def create_skill(
    request: SkillCreate,
    db: Session = Depends(get_db),
) -> dict:
    """Create a new Skill."""
    service = SkillService(db)
    try:
        skill = service.create_skill(request, user_id=None)
        return _make_response(
            {
                "id": skill.id,
                "code": skill.code,
                "name": skill.name,
                "level": skill.level.value,
                "tags": skill.tags or [],
                "status": skill.status.value,
                "draft_snapshot": None,
                "create_user_id": skill.create_user_id,
                "created_at": skill.created_at.isoformat(),
                "updated_at": skill.updated_at.isoformat(),
            }
        )
    except Exception as e:
        if "already exists" in str(e):
            return _make_error_response(ERR_CONFLICT, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.get("", response_model=dict)
async def list_skills(
    status: str | None = Query(None, description="Filter by status"),
    level: str | None = Query(None, description="Filter by level"),
    tags: str | None = Query(None, description="Filter by tags (comma-separated)"),
    search: str | None = Query(None, description="Search by name or code"),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(get_db),
) -> dict:
    """List Skills with filters and pagination."""
    # Parse enums
    status_enum = None
    if status:
        try:
            status_enum = SkillStatus(status)
        except ValueError:
            return _make_error_response(ERR_BAD_REQUEST, f"Invalid status: {status}")

    level_enum = None
    if level:
        try:
            level_enum = SkillLevel(level)
        except ValueError:
            return _make_error_response(ERR_BAD_REQUEST, f"Invalid level: {level}")

    tags_list = None
    if tags:
        tags_list = [t.strip() for t in tags.split(",")]

    service = SkillService(db)
    skills, total = service.list_skills(
        status=status_enum,
        level=level_enum,
        tags=tags_list,
        search=search,
        page=page,
        page_size=page_size,
    )

    items = []
    for skill in skills:
        items.append(
            {
                "id": skill.id,
                "code": skill.code,
                "name": skill.name,
                "level": skill.level.value,
                "tags": skill.tags or [],
                "status": skill.status.value,
                "current_version": None,
                "created_at": skill.created_at.isoformat(),
            }
        )

    return _make_response(
        {
            "total": total,
            "items": items,
            "page": page,
            "page_size": page_size,
        }
    )


@router.get("/{code}", response_model=dict)
async def get_skill(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Get Skill detail."""
    service = SkillService(db)
    try:
        skill = service.get_skill(code)
        return _make_response(_skill_to_detail(skill, getattr(skill, "versions", None)))
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.patch("/{code}", response_model=dict)
async def update_skill(
    code: str = Path(..., description="Skill code"),
    request: SkillUpdate = None,
    db: Session = Depends(get_db),
) -> dict:
    """Update Skill metadata."""
    service = SkillService(db)
    try:
        skill = service.update_skill(code, request)
        return _make_response(
            {
                "id": skill.id,
                "code": skill.code,
                "name": skill.name,
                "level": skill.level.value,
                "tags": skill.tags or [],
                "status": skill.status.value,
                "draft_snapshot": skill.draft_snapshot,
                "create_user_id": skill.create_user_id,
                "created_at": skill.created_at.isoformat(),
                "updated_at": skill.updated_at.isoformat(),
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.delete("/{code}", response_model=dict)
async def delete_skill(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Soft delete a Skill."""
    service = SkillService(db)
    try:
        service.delete_skill(code)
        return _make_response({"success": True})
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        if "active" in str(e).lower():
            return _make_error_response(ERR_BAD_REQUEST, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.post("/{code}/disable", response_model=dict)
async def disable_skill(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Disable a Skill."""
    service = SkillService(db)
    try:
        skill = service.disable_skill(code)
        return _make_response(
            {
                "code": skill.code,
                "status": skill.status.value,
                "disabled_at": skill.updated_at.isoformat(),
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.post("/{code}/enable", response_model=dict)
async def enable_skill(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Enable a Skill."""
    service = SkillService(db)
    try:
        skill = service.enable_skill(code)
        return _make_response(
            {
                "code": skill.code,
                "status": skill.status.value,
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.post("/{code}/publish", response_model=dict)
async def publish_skill(
    code: str = Path(..., description="Skill code"),
    request: SkillVersionCreate = None,
    db: Session = Depends(get_db),
) -> dict:
    """Publish a Skill version."""
    service = SkillService(db)
    try:
        version = service.publish_skill(code, request)
        return _make_response(
            {
                "id": version.id,
                "skill_id": version.skill_id,
                "version": version.version,
                "file_snapshot": version.file_snapshot,
                "comment": version.comment,
                "created_at": version.created_at.isoformat(),
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        if "already exists" in str(e).lower():
            return _make_error_response(ERR_CONFLICT, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.get("/{code}/versions", response_model=dict)
async def get_versions(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Get version history."""
    service = SkillService(db)
    try:
        versions = service.get_versions(code)
        return _make_response(
            {
                "total": len(versions),
                "versions": versions,
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.post("/{code}/rollback", response_model=dict)
async def rollback_skill(
    code: str = Path(..., description="Skill code"),
    request: SkillRollbackRequest = None,
    db: Session = Depends(get_db),
) -> dict:
    """Rollback a Skill to a previous version."""
    service = SkillService(db)
    try:
        result = service.rollback_skill(code, request)
        return _make_response(
            {
                "code": result["code"],
                "draft_snapshot": result["draft_snapshot"],
                "rolled_back_version": result["rolled_back_version"],
            }
        )
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.get("/{code}/files", response_model=dict)
async def get_file_tree(
    code: str = Path(..., description="Skill code"),
    db: Session = Depends(get_db),
) -> dict:
    """Get file tree for a Skill."""
    service = SkillService(db)
    try:
        tree = service.get_file_tree(code)
        return _make_response({"tree": tree})
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.get("/{code}/files/{path:path}", response_model=dict)
async def get_file_content(
    code: str = Path(..., description="Skill code"),
    path: str = Path(..., description="File path"),
    db: Session = Depends(get_db),
) -> dict:
    """Get file content."""
    service = SkillService(db)
    try:
        result = service.get_file_content(code, path)
        return _make_response({"content": result})
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.post("/{code}/files", response_model=dict)
async def create_file(
    code: str = Path(..., description="Skill code"),
    request: FileMetadataCreate = None,
    db: Session = Depends(get_db),
) -> dict:
    """Create a file in a Skill."""
    service = SkillService(db)
    try:
        result = service.create_file(code, request)
        return _make_response(result)
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        if "already exists" in str(e).lower():
            return _make_error_response(ERR_CONFLICT, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.put("/{code}/files/{path:path}", response_model=dict)
async def update_file(
    code: str = Path(..., description="Skill code"),
    path: str = Path(..., description="File path"),
    request: FileMetadataUpdate = None,
    db: Session = Depends(get_db),
) -> dict:
    """Update a file in a Skill."""
    service = SkillService(db)
    try:
        result = service.update_file(code, path, request)
        return _make_response(result)
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))


@router.delete("/{code}/files/{path:path}", response_model=dict)
async def delete_file(
    code: str = Path(..., description="Skill code"),
    path: str = Path(..., description="File path"),
    db: Session = Depends(get_db),
) -> dict:
    """Delete a file from a Skill."""
    service = SkillService(db)
    try:
        service.delete_file(code, path)
        return _make_response({"success": True})
    except Exception as e:
        if "not found" in str(e):
            return _make_error_response(ERR_NOT_FOUND, str(e))
        if "active" in str(e).lower():
            return _make_error_response(ERR_BAD_REQUEST, str(e))
        return _make_error_response(ERR_BAD_REQUEST, str(e))

File: neo/frontend/lib/api/skills.ts
/**
 * Skills API Client
 * Handles all skill-related API calls
 */

import type {
  Skill,
  SkillCreateInput,
  SkillUpdateInput,
  SkillListQuery,
  FileNode,
  FileContentResponse,
  FileCreateInput,
  FileUpdateInput,
  SkillVersion,
  PublishInput,
  RollbackInput,
  ApiResponse,
} from "@/components/skills/skills-types";
import { convertFileTree } from "@/components/skills/skills-types";

// Re-export types
export type {
  Skill,
  SkillCreateInput,
  SkillUpdateInput,
  SkillListQuery,
  FileNode,
  FileContentResponse,
  FileCreateInput,
  FileUpdateInput,
  SkillVersion,
  PublishInput,
  RollbackInput,
  SkillStatus,
  SkillLevel,
  ApiResponse,
  convertFileTree,
} from "@/components/skills/skills-types";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL || "";

// ============================================================
// API Fetch Wrapper
// ============================================================
async function apiFetch<T>(
  endpoint: string,
  options: RequestInit = {},
): Promise<ApiResponse<T>> {
  // Get auth token from store
  const rawToken =
    typeof window !== "undefined" ? localStorage.getItem("neo-auth") : null;
  const parsed = rawToken ? JSON.parse(rawToken) : null;
  const token = parsed?.user?.token ?? parsed?.state?.user?.token ?? null;

  const response = await fetch(`${API_BASE_URL}${endpoint}`, {
    ...options,
    headers: {
      "Content-Type": "application/json",
      ...(token && { Authorization: `Bearer ${token}` }),
      ...options.headers,
    },
    credentials: "include",
  });

  const data = await response.json();

  if (!response.ok) {
    const error = {
      code: data.code || response.status,
      message: data.message || "请求失败",
      detail: data.detail,
    };
    throw error;
  }

  // Check if the API returned an error in the response body
  if (data.code !== 0) {
    const error = {
      code: data.code,
      message: data.message || "请求失败",
      detail: data.detail,
    };
    throw error;
  }

  return data as ApiResponse<T>;
}

// ============================================================
// Skills API
// ============================================================

/**
 * List skills with pagination and filters
 */
export async function listSkills(
  query?: SkillListQuery,
): Promise<{ skills: Skill[]; total: number }> {
  const params = new URLSearchParams();
  if (query?.status) params.set("status", query.status);
  if (query?.level) params.set("level", query.level);
  if (query?.search) params.set("search", query.search);
  if (query?.page) params.set("page", String(query.page));
  if (query?.page_size) params.set("page_size", String(query.page_size));

  const endpoint = `/api/v1/skills${params.toString() ? `?${params}` : ""}`;
  const response = await apiFetch<{
    items: Skill[];
    total: number;
    page: number;
    page_size: number;
  }>(endpoint);

  return {
    skills: response.data?.items || [],
    total: response.data?.total || 0,
  };
}

/**
 * Get skill by code
 */
export async function getSkill(code: string): Promise<Skill> {
  // API returns data directly as Skill object: {code:0, data:{id,code,name,...}}
  const response = await apiFetch<Skill>(`/api/v1/skills/${code}`);
  return response.data!;
}

/**
 * Create a new skill
 */
export async function createSkill(input: SkillCreateInput): Promise<Skill> {
  // API returns data directly as Skill object
  const response = await apiFetch<Skill>("/api/v1/skills", {
    method: "POST",
    body: JSON.stringify(input),
  });
  return response.data!;
}

/**
 * Update skill metadata
 */
export async function updateSkill(
  code: string,
  input: SkillUpdateInput,
): Promise<Skill> {
  // API returns data directly as Skill object
  const response = await apiFetch<Skill>(`/api/v1/skills/${code}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
  return response.data!;
}

/**
 * Delete (soft) skill
 */
export async function deleteSkill(code: string): Promise<void> {
  await apiFetch(`/api/v1/skills/${code}`, {
    method: "DELETE",
  });
}

/**
 * Disable skill
 */
export async function disableSkill(code: string): Promise<Skill> {
  // API returns data directly with {code, status, ...}
  const response = await apiFetch<{ code: string; status: string }>(
    `/api/v1/skills/${code}/disable`,
    {
      method: "POST",
    },
  );
  return response.data as unknown as Skill;
}

/**
 * Enable skill (re-enable disabled skill)
 */
export async function enableSkill(code: string): Promise<Skill> {
  // API returns data directly with {code, status, ...}
  const response = await apiFetch<{ code: string; status: string }>(
    `/api/v1/skills/${code}/enable`,
    {
      method: "POST",
    },
  );
  return response.data as unknown as Skill;
}

// ============================================================
// Files API
// ============================================================

/**
 * Get file tree for a skill
 */
export async function getFileTree(code: string): Promise<FileNode[]> {
  // Define the backend response type
  type BackendFileNode = {
    id: number;
    name: string;
    path: string;
    type: "file" | "directory";
    size?: number;
    children?: BackendFileNode[];
  };
  // API returns data directly with {tree: [...]}
  type FileTreeResponse = { tree: BackendFileNode[] };
  const response = await apiFetch<FileTreeResponse>(
    `/api/v1/skills/${code}/files`,
  );
  // Convert from backend format to frontend format
  return convertFileTree(response.data?.tree || []);
}

/**
 * Get file content
 */
export async function getFileContent(
  code: string,
  path: string,
): Promise<FileContentResponse> {
  // Encode path for URL safety
  const encodedPath = encodeURIComponent(path);
  // API returns data directly with {content: {...}}
  type FileContentWrapper = { content: FileContentResponse };
  const response = await apiFetch<FileContentWrapper>(
    `/api/v1/skills/${code}/files/${encodedPath}`,
  );
  return response.data!.content;
}

/**
 * Create a new file
 */
export async function createFile(
  code: string,
  input: FileCreateInput,
): Promise<{ file_metadata_id: number }> {
  const response = await apiFetch<{ file_metadata_id: number }>(
    `/api/v1/skills/${code}/files`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
  return response.data!;
}

/**
 * Update file content
 */
export async function updateFile(
  code: string,
  path: string,
  input: FileUpdateInput,
): Promise<void> {
  const encodedPath = encodeURIComponent(path);
  await apiFetch(`/api/v1/skills/${code}/files/${encodedPath}`, {
    method: "PUT",
    body: JSON.stringify(input),
  });
}

/**
 * Delete file
 */
export async function deleteFile(code: string, path: string): Promise<void> {
  const encodedPath = encodeURIComponent(path);
  await apiFetch(`/api/v1/skills/${code}/files/${encodedPath}`, {
    method: "DELETE",
  });
}

// ============================================================
// Version API
// ============================================================

/**
 * Get version history
 */
export async function getVersions(code: string): Promise<SkillVersion[]> {
  // API returns data directly with {versions: [...]}
  type VersionsResponse = { versions: SkillVersion[]; total: number };
  const response = await apiFetch<VersionsResponse>(
    `/api/v1/skills/${code}/versions`,
  );
  return response.data?.versions || [];
}

/**
 * Publish skill (create new version)
 */
export async function publishSkill(
  code: string,
  input: PublishInput,
): Promise<SkillVersion> {
  // API returns data directly as version object: {id, skill_id, version, ...}
  type PublishResponse = {
    id: number;
    skill_id: number;
    version: string;
    file_snapshot: unknown[];
    comment: string;
    created_at: string;
  };
  const response = await apiFetch<PublishResponse>(
    `/api/v1/skills/${code}/publish`,
    {
      method: "POST",
      body: JSON.stringify(input),
    },
  );
  return response.data as unknown as SkillVersion;
}

/**
 * Rollback to previous version
 */
export async function rollbackSkill(
  code: string,
  input: RollbackInput,
): Promise<{ skill: Skill }> {
  // API returns data directly: {code, draft_snapshot, rolled_back_version}
  type RollbackResponse = {
    code: string;
    draft_snapshot: unknown[];
    rolled_back_version: string;
  };
  await apiFetch<RollbackResponse>(`/api/v1/skills/${code}/rollback`, {
    method: "POST",
    body: JSON.stringify(input),
  });
  // Return empty - caller should reload skill data
  return { skill: {} as Skill };
}
